Address-Book.go

```diff
@@ -0,0 +1,464 @@
+	} else {
+		// If the login succeeded
+		SaveUserSession(context,username,w,r)
+		http.Redirect(w, r, "/userpage", 301)
+		return
+	}
+}
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+func (usersession  *UserSession) Register(context *GlobalData,username string ,password string , w http.ResponseWriter, r *http.Request ){
+
+
+
+	err :=CheckUsernameExists(context , username)
+	switch {
+	case err == nil:
+		// Username is not available
+		usersession.err="Please choose a different username"
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+		return
+	case err == gocql.ErrNotFound:
+		// Username is available
+		hashedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+		if err != nil {
+			fmt.Println("Couldn't Incrypt")
+			usersession.err="This Password is Not premitted"
+			http.Error(w,err.Error(),http.StatusInternalServerError)
+			return
+		}
+
+		err = InsertUser(context,username,hashedPassword)
+		if err != nil {
+			http.Error(w,err.Error(),http.StatusInternalServerError)
+			return
+		}
+		SaveUserSession(context,username,w,r)
+		http.Redirect(w, r, "/userpage", http.StatusFound)
+		return
+	case err != nil:
+		//Database Error
+		http.Error(w, "Server error, unable to create your account.", 500)
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+		return
+	default:
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+	}
+
+}
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+func (usersession  *UserSession) Login(context *GlobalData, w http.ResponseWriter, r *http.Request){
+
+	username := r.FormValue("username")
+	password := r.FormValue("password")
+
+	//Inputs Validation
+	if len(username) > 50 {
+		usersession.err="Username is too long"
+		http.Redirect(w,r,"/home",301)
+		return
+	}
+	if len(password) > 120 {
+
+		usersession.err="Password is too long"
+		http.Redirect(w,r,"/home",301)
+		return
+	}
+	if username == "" {
+
+		usersession.err="Please Enter a Username"
+		http.Redirect(w,r,"/home",301)
+		return
+	}
+	if password == "" {
+		usersession.err="Please Enter a password"
+		http.Redirect(w,r,"/home",301)
+		return
+	}
+	if validateEmail(username) == false {
+		usersession.err="Please Enter a valid Username"
+		http.Redirect(w,r,"/home",301)
+		return
+	}
+
+	if r.FormValue("register")!="" {
+		usersession.Register(context , username ,password ,w,r)
+	}else if r.FormValue("login")!="" {
+		usersession.SignIn(context , username ,password ,w,r)
+	}
+
+
+}
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+func Write (w http.ResponseWriter, r *http.Request , next http.HandlerFunc){
+//fmt.Println("I am in middleWare")
+	next(w,r)
+}
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+func QueryUser(context *GlobalData , username string) (string,error){
```

Comment:
This function belong to which model ?? 
try to remember Modeling concepts we talked .